### ISSUE TITLE
Unused lambda capture source ranges are slightly off

### ISSUE THREAD

**Issue by tbaederr** [llvm/llvm-project#106445]
See: https://godbolt.org/z/MKjzsYb9b

```c++
int main() {

    int a = 0, b = 0;

    auto F = [&a, &b]() {

    };
}
```
The output is:
```console
<source>:7:16: warning: lambda capture 'a' is not used [-Wunused-lambda-capture]
    7 |     auto F = [&a, &b]() {
      |               ~^~
<source>:7:20: warning: lambda capture 'b' is not used [-Wunused-lambda-capture]
    7 |     auto F = [&a, &b]() {
      |                 ~~~^
<source>:7:10: warning: unused variable 'F' [-Wunused-variable]
    7 |     auto F = [&a, &b]() {
      |          ^
```

1) The source range for `&a` includes the comma after it
2) The source range for `&b` includes both the comma and the whitespace between that and the capture

The second point is especially fun when inserting some useless whitespace:
```c++
    auto F = [&a,         
    
     &b]() {

    };
```

**Comment by shafik** [comment]
Note the source range for `a` does not include the comma after it if you have whitespace: https://godbolt.org/z/sGEYPaThs

**Comment by tbaederr** [comment]
Right, those aren't just ranges, but delete ranges from the fixme. In that case, I'm not sure what to do. :upside_down_face:

**Comment by jhendle2** [comment]
For https://godbolt.org/z/8zdKPxY6K
```c++
int main() {

    int a = 0, b = 0;

    auto F = [&a, &b]() {

    };
}
```
The output is:
```
<source>:11:16: warning: lambda capture 'a' is not used [-Wunused-lambda-capture]
   11 |     auto F = [&a, &b]() {
      |               ~^~
<source>:11:20: warning: lambda capture 'b' is not used [-Wunused-lambda-capture]
   11 |     auto F = [&a, &b]() {
      |                 ~~~^
<source>:11:10: warning: unused variable 'F' [-Wunused-variable]
   11 |     auto F = [&a, &b]() {
      |          ^
```

So, I assume the third warning output should be:
```
<source>:11:20: warning: lambda capture 'b' is not used [-Wunused-lambda-capture]
   11 |     auto F = [&a, &b]() {
      |                   ~^
```

Because this matches the behavior for a lambda with a single unsued capture?
```
<source>:11:16: warning: lambda capture 'a' is not used [-Wunused-lambda-capture]
   11 |     auto F = [&a]() {
      |               ~^
```

**Comment by tbaederr** [comment]
Kinda, yes. But the output is already correct as it is, when we consider that the source ranges aren't supposed to just highlight the unused parameter. They are fix-it hints that will _remove_ the underlined text when applied.

**Comment by jhendle2** [comment]
Hi, can this ticket be assigned to me?

**Comment by charan-003** [comment]
Hii, can i please work on this?
can this be assigned to me

**Comment by charan-003** [comment]
@shafik @tbaederr @jhendle2 can you assign this to me

**Comment by jhendle2** [comment]
Hi Charan - I have been working on this ticket separately and am wrapping up my changes this week.

**Comment by erichkeane** [comment]
@jhendle2 : I don't see a patch from you, but please see #117953, which might fix this/is under review.

**Comment by zwuis** [comment]
I believe the root cause of this issue is that the character "~" is hard to show that this is a fix-it hint.

How can you see some code marked by "~" and understand this is a fix-it hint suggesting removing this part of code? No, you can't. This is also why this issue was created.

In the short term (to just fix this issue), my ideas are:

1. Just remove the fix-it hint. This is the easiest way.
2. Add a note in diagnostic messages, and move the fix-it hint to the note. E.g.

```
<source>:7:20: warning: lambda capture 'b' is not used [-Wunused-lambda-capture]
7 | auto F = [&a, &b]() {
| ^
<source>:7:20: note: you can remove this part of code
7 | auto F = [&a, &b]() {
| ~~~^
```

In the long term, if necessary, we need to reconsider how to show fix-it hint in command line output. E.g. changing to diff style:

```diff
@@ -4,7 +4,7 @@
int a = 0, b = 0;
- auto F = [&a, &b]() {
+ auto F = [&a]() {
};
}
```

**Comment by ojhunt** [comment]
I've put up a PR to fix this bug from a technical standpoint: the fixit ranges produced are now consistently correct, in that they highlight the full range of the required FixIt, but this doesn't solve the issue @tbaederr has in that the highlight range is the fixit deletion range, not simply the capture declaration.

Trying to fix that is more complicated as the FixIt range definitionally overlaps the actual capture expression, which means any attempt to point to just the capture itself gets subsumed by the FixIt itself.

This might be resolvable by some CLI UI changes for FixIts, maybe to make deletion vs insertion more clear, etc. Separating the rendering of the fixit may also make it possible to independently point to the actual capture expression.

One issue I do foresee there is that I think the capture range _may_ include more text than just the capture expression, but I'm not sure just how precise it is.

I also just realized that one of my tests involves macros, and I just realized that the improved capture ranges will actually result in no FixIt being present (due to the inability to safely modify a macro), but as a result there's probably no longer anything pointing to the sadness capture. I'll see if I can improve the warning in cases like that, and then hopefully we'll automatically resolve capture range vs fixit range if we do change the rendering in future.

(Just to clarify: I mean the cli rendering of fixits, not I'm not considering any higher level tooling)

**Comment by EugeneZelenko** [comment]
Pull request is not merged yet.

**Comment by ojhunt** [comment]
@EugeneZelenko  Argh! Thanks for catching that muppetry! GitHub UI and I do not work well together -- I even noted the terrible UI (took a screenshot to complain to a friend), and then clicked the wrong button anyway.
